plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -935,9 +941,81 @@ public Optional<ConnectorTableHandle> applyDelete(ConnectorSession session, Conn
     }
 
     @Override
-    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
+    public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        if (table.getFormatVersion() < 2) {
+            throw new TrinoException(NOT_SUPPORTED, "Iceberg table updates require at least format version 2");
+        }
+        String deleteMode = table.getStorageProperties().get(DELETE_MODE);
+        if (!"merge-on-read".equals(deleteMode)) {
+            String errorMessage = "Row level deletes are only supported for tables with '" + DELETE_MODE + "' set to 'merge-on-read'.";
+            if (deleteMode != null) {
+                errorMessage += " '" + deleteMode + "' is not supported.";
+            }
+            throw new TrinoException(NOT_SUPPORTED, errorMessage);
+        }
+        verify(transaction == null, "transaction already set");
+        transaction = catalog.loadTable(session, table.getSchemaTableName()).newTransaction();
+        return table.withRetryMode(retryMode);
+    }
+
+    @Override
+    public void finishDelete(ConnectorSession session, ConnectorTableHandle tableHandle, Collection<Slice> fragments)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) tableHandle;
+        Table icebergTable = transaction.table();
+        FileFormat format = getFileFormat(icebergTable).toIceberg();
+
+        List<CommitTaskData> commitTasks = fragments.stream()
+                .map(slice -> commitTaskCodec.fromJson(slice.getBytes()))
+                .collect(toImmutableList());
+
+        Type[] partitionColumnTypes = icebergTable.spec().fields().stream()
+                .map(field -> field.transform().getResultType(icebergTable.schema().findType(field.sourceId())))
+                .toArray(Type[]::new);
+
+        RowDelta rowDelta = transaction.newRowDelta();
+        table.getSnapshotId().map(icebergTable::snapshot).ifPresent(s -> rowDelta.validateFromSnapshot(s.snapshotId()));
+        ImmutableSet.Builder<String> writtenFiles = ImmutableSet.builder();
+        ImmutableSet.Builder<String> referencedDataFiles = ImmutableSet.builder();
+        for (CommitTaskData task : commitTasks) {
+            if (task.getContent() != FileContent.POSITION_DELETES) {
+                throw new TrinoException(GENERIC_INTERNAL_ERROR, "Iceberg finishDelete called with commit task that was not a position delete file");
+            }
+            FileMetadata.Builder deleteBuilder = FileMetadata.deleteFileBuilder(icebergTable.spec())
```

Comment:
@rdblue maybe you can help clarify one thing for me. In the case of a position delete where the data file being deleted from has an old PartitionSpec, does  `FileMetadata.deleteFileBuilder` expect the spec and partition information used by the data file at the time it was written, or the current table version?

Reply:
Delete files must have the same spec and partition as the data files they are applied to.

There isn't a "current" table spec, there is only the default spec to be used when there isn't a constraint that requires a different one -- either a user explicitly writing with a spec or cases like this, where the spec is determined by the data files you're trying to delete rows in.

Reply:
Good to know, thanks!

Reply:
I came across one possible bug in the Spark/Core library while writing a test for this. https://github.com/apache/iceberg/issues/4563